plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -95,8 +101,113 @@ public void start() {
         showAllSecurityWarnings();
         bundledPlugins();
         installedPlugins();
-        downloadPlugins(plugins);
-        outputFailedPlugins();
+
+        allPluginsAndDependencies = findPluginsAndDependencies(cfg.getPlugins());
+        pluginsToBeDownloaded = findPluginsToDownload(allPluginsAndDependencies);
+        effectivePlugins = findEffectivePlugins(pluginsToBeDownloaded);
+
+        listPlugins();
+        showSpecificSecurityWarnings(pluginsToBeDownloaded);
+        showAvailableUpdates(pluginsToBeDownloaded);
+
+        if (!cfg.isShowPluginsToBeDownloaded() && !cfg.isShowAllWarnings() && !cfg.isShowWarnings() &&
+                !cfg.isShowAvailableUpdates()) {
+            downloadPlugins(pluginsToBeDownloaded);
+            outputFailedPlugins();
+        }
+    }
+
+    /**
+     * Compares the list of all requested plugins to the currently installed plugins to determine the final list of
+     * plugins that will be downloaded
+     * @param requestedPlugins list of all requested plugins, determined from the highest required versions of the
+     *                         initial user requested plugins and their recursive dependencies
+     * @return
+     */
+    public List<Plugin> findPluginsToDownload(Map<String, Plugin> requestedPlugins) {
+        List<Plugin> pluginsToDownload = new ArrayList<>();
+        VersionNumber installedVersion = null;
+        for (Map.Entry<String, Plugin> requestedPlugin : requestedPlugins.entrySet()) {
+            String pluginName = requestedPlugin.getKey();
+            Plugin plugin = requestedPlugin.getValue();
+            if (installedPluginVersions.containsKey(pluginName)) {
+                installedVersion = installedPluginVersions.get(pluginName);
+            } else if (bundledPluginVersions.containsKey(pluginName)) {
+                installedVersion = bundledPluginVersions.get(pluginName);
+            }
+            if (installedVersion == null) {
+                pluginsToDownload.add(plugin);
+            } else if (installedVersion.compareTo(plugin.getVersion()) < 0) {
+                if (isVerbose) {
+                    System.out.println("Installed version (" + installedVersion + ") of " + pluginName +
+                            " is less than minimum required version of " + plugin.getVersion() +
+                            ", bundled plugin will be upgraded");
+                }
+                pluginsToDownload.add(plugin);
+            }
+        }
+        return pluginsToDownload;
+    }
+
+    /**
+     * Finds the final set of plugins that have been
+     * @param pluginsToBeDownloaded
+     * @return
+     */
+    public Map<String, Plugin> findEffectivePlugins(List<Plugin> pluginsToBeDownloaded) {
+        Map<String, Plugin> effectivePlugins = new HashMap<>();
+        for (Plugin plugin : pluginsToBeDownloaded) {
```

Comment:
You could probably use streams here as well like
`
List<String> copy = list.stream()
  .collect(Collectors.toList());
`
and then keep passing the stream on to add the other plugins.